src/rviz/default_plugin/map_display.cpp

```diff
@@ -59,9 +59,162 @@
+  ss3 << "NodeObject" << node_count++;
+
+  scene_node_ = parent_->scene_node_->createChildSceneNode(ss3.str());
+  scene_node_->attachObject( manual_object_ );
+
+  manual_object_->begin(material_->getName(), Ogre::RenderOperation::OT_TRIANGLE_LIST);
+  {
+    // First triangle
+    {
+      // Bottom left
+      manual_object_->position( 0.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top right
+      manual_object_->position( 1.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top left
+      manual_object_->position( 0.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+    }
+
+    // Second triangle
+    {
+      // Bottom left
+      manual_object_->position( 0.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(0.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Bottom right
+      manual_object_->position( 1.0f, 0.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 0.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+
+      // Top right
+      manual_object_->position( 1.0f, 1.0f, 0.0f );
+      manual_object_->textureCoord(1.0f, 1.0f);
+      manual_object_->normal( 0.0f, 0.0f, 1.0f );
+    }
+  }
+  manual_object_->end();
+
+  scene_node_->setPosition( x*resolution, y*resolution, 0 );
+  scene_node_->setScale( width*resolution, height*resolution, 1.0 );
+
+  if( parent_->draw_under_property_->getValue().toBool() )
+  {
+    manual_object_->setRenderQueueGroup(Ogre::RENDER_QUEUE_4);
+  }
+
+  // don't show map until the plugin is actually enabled
+  manual_object_->setVisible( false );
+
+
+}
+
+Swatch::~Swatch()
+{
+    delete manual_object_;
+}
+
+void Swatch::updateAlpha(const Ogre::SceneBlendType sceneBlending, bool depthWrite, AlphaSetter* alpha_setter)
+{
+  Ogre::Pass* pass = material_->getTechnique( 0 )->getPass( 0 );
+  Ogre::TextureUnitState* tex_unit = NULL;
+
+  material_->setSceneBlending( sceneBlending ); 
+  material_->setDepthWriteEnabled( depthWrite );
+  if( manual_object_ )
+  {
+    manual_object_->visitRenderables( alpha_setter );
+  }
+}
+
+void Swatch::updateData()
+{
+  unsigned int pixels_size = width_ * height_;
+  unsigned char* pixels = new unsigned char[pixels_size];
+  memset(pixels, 255, pixels_size);
+  unsigned char* ptr = pixels;
+  int N = parent_->current_map_.data.size();
+  unsigned int fw = parent_->current_map_.info.width;
+
+  for(unsigned int yy=y_; yy<y_+height_;yy++){
+    int index = yy * fw + x_;
+    int pixels_to_copy = std::min((int)width_, N-index);
+    memcpy(ptr, &parent_->current_map_.data[ index ], pixels_to_copy); 
+    ptr+=pixels_to_copy;
+    if(index+pixels_to_copy>=N) break;
+  }
+
+  Ogre::DataStreamPtr pixel_stream;
+  pixel_stream.bind( new Ogre::MemoryDataStream( pixels, pixels_size ));
```

Comment:
This looks ok, since `pixel_stream` is a shared pointer of Ogre's own making. The memory for the `MemoryDataStream` should be reclaimed correctly.